Add participant languages

dodumosu opened the pull request:
This PR adds support for responding to text messages in a preset language for a participant (see nditech/apollo-issues#176).

In order to set this up:
- the deployment primary and other languages must be set up in the admin;
- individual participant languages should be set by importing in the specified column a two-letter language code that matches one of the supported languages, OR
- edit a participant and set the language in the edit participant screen

Participants that have no language set default to the default language for the deployment, and the same occurs for messages that cannot load a valid participant.
![ifNYgQSFGt](https://user-images.githubusercontent.com/1408662/180489045-09df1b6e-f4aa-4f04-bceb-d3d5884876cb.png)

![eM0Z5LX1ai](https://user-images.githubusercontent.com/1408662/180489488-e539d348-4e11-4658-baa0-b87a8b1d3a6a.png)

In the screenshots above participant 2001 has their language set to French, 2002 has theirs set to English, and 2003 has no language set

dodumosu committed:
feat: add locale switcher

dodumosu committed:
feat: add locale field to participant

dodumosu committed:
feat: enable editing for participant locale

dodumosu committed:
feat: add participant language code import

dodumosu committed:
chore: force locale code to lowercase

dodumosu committed:
fix: add the text/plain mimetype to CSV types

dodumosu committed:
feat: change locale based on participant

dodumosu commented:
The PR is this large because the language catalogs have been rebuilt.

takinbo commented:
Is the catalog containing any new translation strings? If not, I would suggest reviewing the PR and excluding any files that didn't undergo any meaningful change. I'll preview to review a smaller PR.

dodumosu commented:
It doesn't contain updated strings, but the number of lines in the source code have changed, and the translations won't work without being rebuilt. I can remove the commit that changes the translations, but you will need to rebuild them locally to test, for example.

dodumosu committed:
feat: various changes

- add participant locale to API login payload
- add locale serialization for participants

takinbo commented on apollo/participants/models.py:
```diff
@@ -52,7 +52,8 @@ def get_import_fields(self):
             'email': _('Email'),
-            'password': _('Password')
+            'password': _('Password'),
+            'locale': _('Language code'),
```
Can we name this simply Language or we call it Locale?

takinbo commented on apollo/frontend/forms.py:
```diff
@@ -90,6 +90,9 @@ def generate_participant_edit_form(participant, data=None):
     attributes['phone'] = StringField(_('Phone'))
     attributes['password'] = StringField(_('Password'))
+    attributes['locale'] = SelectField(
+        _('Language'),
```
"Locale"

takinbo commented on apollo/messaging/helpers.py:
```diff
@@ -1,30 +1,46 @@
-    (prefix, participant_id, exclamation, form_serial, responses, comment) = \
-        parse_text(message['text'])
-    if (prefix and participant_id and responses):
+    (
+        prefix,
+        participant_id,
+        exclamation,
+        form_serial,
+        responses,
+        comment,
+    ) = parse_text(message["text"])
+    if prefix and participant_id and responses:
```
We need to come up with a consistent code writing style and stick to it. It should overtime lead to less and less code rewrites like this. This is essentially a cosmetic change.

dodumosu replied:
we could choose a formatter and rules and use that?

takinbo reviewed (changes requested):
I'll like to see unit tests written for this feature.

dodumosu committed:
feat: add tests for translation context switch

dodumosu committed:
chore: change labels

dodumosu committed:
chore: refactor tests

dodumosu committed:
chore: remove deprecation warnings

dodumosu committed:
chore: stylistic changes

dodumosu committed:
fix: skip loading prometheus during tests

this commit allows tests pass on CircleCI by skipping loading
prometheus. on nose (which is configured for CircleCI), running
tests with the full app loaded (including prometheus) throws up
errors. using pytest as the test runner, those errors don't show
up.

dodumosu committed:
fix: compile language catalogs on CircleCI

takinbo commented on apollo/__init__.py:
```diff
@@ -95,7 +95,10 @@ def delay_flask_security_mail(msg):
     init_admin(admin, app)
 
-    monitor(app)
+    if not app.config.get('TESTING'):
+        # there's a known issue when prometheus is loaded
+        # during testing.
+        monitor(app)
```
Is this still an issue?

dodumosu replied:
since we've changed from using nose to pytest, no

takinbo commented on apollo/testutils/factory.py:
```diff
@@ -27,3 +28,17 @@ def create_test_app():
     return app
+
+
+def create_full_test_app():
+    test_settings = {
+        'SQLALCHEMY_DATABASE_URI': settings.TEST_DATABASE_URL,
+        'TESTING': True,
+    }
+
+    testConfig = TestConfig()
+    testConfig.update(**test_settings)
+
+    app = create_full_app(testConfig)
+
+    return app
```
I don't understand the need for this. These variables are configured automatically in the test environment why create another method for this?

dodumosu replied:
it's not for the variables, though. this is related to your comment on nditech/apollo#866. the app created using `create_app()` from `apollo/factory.py` doesn't have flask-security set up, while `create_app()` from `apollo/__init__.py` does have flask-security set up properly using the user datastore.

dodumosu replied:
![WindowsTerminal_8Mp3WUtkJs](https://user-images.githubusercontent.com/1408662/205835507-d7d0b090-52f3-41d5-ae54-6d04896a46d5.png)
An example of what happens when `create_app()` from `apollo/factory.py` is used instead of from the package module, because flask-security wasn't set up

dodumosu committed:
chore: remove exclusion for nose

dodumosu committed:
chore: remove setting redefinition

takinbo commented on apollo/testutils/factory.py:
```diff
@@ -27,3 +27,16 @@ def create_test_app():
+
+    app = create_full_app(testConfig)
+
+    return app
```
Do we still need this?

dodumosu committed:
chore: reorder migrations

dodumosu committed:
chore: remove unnecessary code

takinbo commented on apollo/frontend/forms.py:
```diff
@@ -90,6 +90,9 @@ def generate_participant_edit_form(participant, data=None):
         _('Partner'), choices=_make_choices(partner_choices))
     attributes['phone'] = StringField(_('Phone'))
     attributes['password'] = StringField(_('Password'))
+    attributes['locale'] = SelectField(
+        _('Locale'),
+        choices=_make_choices(languages.items(), _('Unspecified')))
```
It could be a little confusing for users to use the term "Locale" in referring to the preferred language. On the backend, we can use locale but we should use "Preferred Language" on the user-facing side.

takinbo reviewed (changes requested):
Change the user-facing language for "Locale" to "Preferred Language"

dodumosu committed:
chore: change label

takinbo commented on apollo/participants/models.py:
```diff
@@ -52,7 +52,8 @@ def get_import_fields(self):
             'email': _('Email'),
-            'password': _('Password')
+            'password': _('Password'),
+            'locale': _('Locale'),
```
Let's use `_('Preferred Language')` instead.

takinbo commented on apollo/frontend/forms.py:
```diff
@@ -90,6 +90,9 @@ def generate_participant_edit_form(participant, data=None):
     attributes['password'] = StringField(_('Password'))
+    attributes['locale'] = SelectField(
+        _('Preferred Language'),
+        choices=_make_choices(languages.items(), _('Unspecified')))
```
Instead of using `Unspecified` let's change it to `Default` so it's not confused as to whether it has to be specified.

takinbo reviewed (changes requested):
Besides the individual comments, I noticed that responses are returned in the configured locale if the text message is valid (even if the responses are not); for invalid messages, would it be possible to attempt a lookup of the phone number to identify the user? This might offer a better user experience. If the observer can be identified by the id in the text message, fine if not, attempt the identification based on the sender.

dodumosu committed:
chore: modify labels

dodumosu committed:
chore: update docstring

dodumosu committed:
feat: switch text response language

takinbo reviewed (approved):
utACK
